etk-asm/src/asm.rs

```diff
@@ -244,56 +233,102 @@ impl Assembler {
-    /// may remain.
-    pub fn finish(self) -> Result<(), Error> {
-        if let Some(undef) = self.pending.front() {
-            return match undef {
-                RawOp::Op(AbstractOp::Macro(invc)) => error::UndeclaredInstructionMacro {
-                    name: invc.name.clone(),
-                }
-                .fail(),
-                RawOp::Op(op) => {
-                    match op
-                        .clone()
-                        .concretize((&self.declared_labels, &self.declared_macros).into())
-                    {
-                        Ok(_) => unreachable!(),
-                        Err(ops::Error::ContextIncomplete {
-                            source: expression::Error::UnknownMacro { name, .. },
-                            ..
-                        }) => error::UndeclaredExpressionMacro { name }.fail(),
-                        Err(ops::Error::ContextIncomplete {
-                            source: expression::Error::UnknownLabel { .. },
-                            ..
-                        }) => {
-                            let labels = op.expr().unwrap().labels(&self.declared_macros).unwrap();
-                            let declared = self.declared_labels.into_keys().collect();
-                            let invoked: HashSet<_> = labels.into_iter().collect();
-                            let missing = invoked
-                                .difference(&declared)
-                                .cloned()
-                                .collect::<Vec<String>>();
-                            error::UndeclaredLabels { labels: missing }.fail()
-                        }
-                        _ => unreachable!(),
-                    }
-                }
-                // bug: if a variable is used when it isn't available, e.g. push1 $size
-                _ => unreachable!(),
-            };
+    /// Create a new `Assembler` with a known length.
+    pub fn new_internal(actual_len: usize) -> Self {
+        Self {
+            concrete_len: actual_len,
+            ..Self::default()
         }
+    }
+
+    /// Get the number of bytes that can be collected with [`Assembler::take`].
+    pub fn concrete_len(&self) -> usize {
+        self.concrete_len
+    }
 
-        if !self.ready.is_empty() {
-            panic!("not all assembled bytecode has been taken");
+    /// Inspect the macros in a series of [`RawOp`] and declare them in the assembler.
+    pub fn inspect_macros<I, O>(&mut self, nodes: I) -> Result<(), Error>
+    where
+        I: IntoIterator<Item = O>,
+        O: Into<RawOp>,
+    {
+        for op in nodes {
+            let op = op.into();
+            if let RawOp::Op(AbstractOp::MacroDefinition(_)) = op {
+                self.declare_macro(op)?
+            }
         }
 
         Ok(())
     }
 
     /// Collect any assembled instructions that are ready to be output.
     pub fn take(&mut self) -> Vec<u8> {
```

Comment:
Did you consider changing the API to just an `assemble` function? I don't think we need the `push_all`/`take` phases if we're only considering the whole program in one chunk.

Reply:
I fully agree with this. I made the change from push_all to assemble and now the whole process is done through a single call. Even the creation of a new assembler to resolve scopes is now done internally, which makes much more sense.

Thanks!